File: lintcode/python/0070_binary_tree_level_order_traversal_II.py
#!/usr/bin/python -t

# bfs, queue and tree traversal solution

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

from collections import deque

class Solution:
    """
    @param root: A tree
    @return: buttom-up level order a list of lists of integer
    """
    def levelOrderBottom(self, root):
        # write your code here
        ret = []
        if root == None:
            return []
            
        q = deque([root])
        
        while q:
            level = []
            
            for i in range(len(q)):
                node = q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
                
                level.append(node.val)
                #print level
            ret.append(level)
        
        ret.reverse()
        return ret

File: lintcode/golang/0070_binary_tree_level_order_traversal_II.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: A tree
 * @return: buttom-up level order a list of lists of integer
 */
func levelOrderBottom (root *TreeNode) [][]int {
    // write your code here
    if root == nil {
        return [][]int{}
    }
    
    q := []*TreeNode{root}
    var ret [][]int
    
    for len(q) != 0 {
        level := []*TreeNode{}
        var path []int
        for _, node := range q {
            path = append(path, node.Val)
            if node.Left != nil {
                level = append(level, node.Left)
            }
            if node.Right != nil {
                level = append(level, node.Right)
            }
        }
        q = level
        ret = append(ret, path)
    }
    
    ret = reverse(ret)
    
    return ret
}

func reverse(s [][]int) [][]int {
    //a := make([][]int, len(s))
    //copy(a, s)
    a := s

    for i := len(a)/2 - 1; i >= 0; i-- {
        opp := len(a) - 1 - i
        a[i], a[opp] = a[opp], a[i]
    }

    return a
}
